hatchery/gen3licenseusermaps.go

```diff
@@ -0,0 +1,369 @@
+
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Gen3LicenseUserMap struct {
+	ItemId             string `json:"itemId"`
+	Environment        string `json:"environment"`
+	LicenseType        string `json:"licenseType"`
+	IsActive           string `json:"isActive"`
+	UserId             string `json:"userId"`
+	LicenseId          int    `json:"licenseId"`
+	FirstUsedTimestamp int    `json:"firstUsedTimestamp"`
+	LastUsedTimestamp  int    `json:"lastUsedTimestamp"`
+}
+
+var initializeDbConfig = func() *DbConfig {
+	// Create a new dynamoDB client
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
+			Region: aws.String("us-east-1"),
+			// Use this endpoint for running locally
+			// Endpoint: aws.String("http://localhost:8000"),
+		},
+	}))
+	return &DbConfig{
+		DynamoDb: dynamodb.New(sess),
+	}
+}
+
+var validateContainerLicenseInfo = func(containerName string, licenseInfo LicenseInfo) bool {
+
+	ok := true
+	if !licenseInfo.Enabled {
+		fmt.Printf("Warning: License is not enabled for container %s\n", containerName)
+		ok = false
+	}
+	if licenseInfo.LicenseType == "" {
+		fmt.Printf("Error in container config. Empty LicenseType for container %s\n", containerName)
+		ok = false
+	}
+	if licenseInfo.MaxLicenseIds == 0 {
+		fmt.Printf("Error in container config. Empty or 0 MaxLicenseIds for container %s\n", containerName)
+		ok = false
+	}
+	if licenseInfo.G3autoName == "" {
+		fmt.Printf("Error in container config. Empty G3autoName for container %s\n", containerName)
+		ok = false
+	}
+	if licenseInfo.G3autoKey == "" {
+		fmt.Printf("Error in container config. Empty G3autoKey for container %s\n", containerName)
+		ok = false
+	}
+	if licenseInfo.FilePath == "" {
+		fmt.Printf("Error in container config. Empty FilePath for container %s\n", containerName)
+		ok = false
+	}
+	if licenseInfo.WorkspaceFlavor == "" {
+		fmt.Printf("Error in container config. Empty WorkspaceFlavor for container %s\n", containerName)
+		ok = false
+	}
+	return ok
+}
+
+var getActiveGen3LicenseUserMaps = func(dbconfig *DbConfig, container Container) (gen3LicenseUserMaps *[]Gen3LicenseUserMap, err error) {
+	// Query the table to get all active gen3 license user map items
+
+	targetEnvironment := os.Getenv("GEN3_ENDPOINT")
+	validateContainerLicenseInfo(container.Name, container.License)
+
+	// Query on global secondary index and filter by license type (eg. "STATA")
+	keyEx1 := expression.Key("environment").Equal(expression.Value(aws.String(targetEnvironment)))
+	keyEx2 := expression.Key("isActive").Equal(expression.Value("True"))
+	filt := expression.Name("licenseType").Equal(expression.Value(container.License.LicenseType))
+	expr, err := expression.NewBuilder().WithKeyCondition(expression.KeyAnd(keyEx1, keyEx2)).WithFilter(filt).Build()
+	if err != nil {
+		Config.Logger.Printf("Error in building expression for query: %s", err)
+		return nil, err
+	}
+	res, err := dbconfig.DynamoDb.Query(&dynamodb.QueryInput{
```

Comment:
the query function can return a paginated result, you need to add a loop to get all the items.
see `dynamodb.QueryOutput.LastEvaluatedKey` [here](https://docs.aws.amazon.com/sdk-for-go/api/service/dynamodb/#QueryOutput) and `dynamodb.QueryInput.ExclusiveStartKey` [here](https://docs.aws.amazon.com/sdk-for-go/api/service/dynamodb/#QueryInput).
Here's an example i just wrote for a different hatchery feature https://github.com/uc-cdis/hatchery/blob/f68e5b11c08de9cf997665c337f8617111d0f447/hatchery/helpers.go#L243-L264

Reply:
Ah, nice catch, thank you! The query has moved into a function. It now utilizes the `LastEvaluatedKey` and `ExclusiveStartKey`.